chalk-ir/src/zip.rs

```diff
@@ -335,21 +472,220 @@ impl<I: Interner> Zip<I> for VariableKind<I> {
 }
 
 impl<I: Interner> Zip<I> for GenericArg<I> {
-    fn zip_with<'i, Z: Zipper<'i, I>>(zipper: &mut Z, a: &Self, b: &Self) -> Fallible<()>
+    fn zip_with<'i, Z: Zipper<'i, I>>(
+        zipper: &mut Z,
+        variance: Variance,
+        a: &Self,
+        b: &Self,
+    ) -> Fallible<()>
     where
         I: 'i,
     {
         let interner = zipper.interner();
-        Zip::zip_with(zipper, a.data(interner), b.data(interner))
+        Zip::zip_with(zipper, variance, a.data(interner), b.data(interner))
     }
 }
 
 impl<I: Interner> Zip<I> for ProgramClause<I> {
-    fn zip_with<'i, Z: Zipper<'i, I>>(zipper: &mut Z, a: &Self, b: &Self) -> Fallible<()>
+    fn zip_with<'i, Z: Zipper<'i, I>>(
+        zipper: &mut Z,
+        variance: Variance,
+        a: &Self,
+        b: &Self,
+    ) -> Fallible<()>
+    where
+        I: 'i,
+    {
+        let interner = zipper.interner();
+        Zip::zip_with(zipper, variance, a.data(interner), b.data(interner))
+    }
+}
+
+impl<I: Interner> Zip<I> for TraitRef<I> {
+    fn zip_with<'i, Z: Zipper<'i, I>>(
+        zipper: &mut Z,
+        variance: Variance,
+        a: &Self,
+        b: &Self,
+    ) -> Fallible<()>
+    where
+        I: 'i,
+    {
+        let interner = zipper.interner();
+        Zip::zip_with(zipper, variance, &a.trait_id, &b.trait_id)?;
```

Comment:
This should probably just be `Invariant`

Reply:
In addition to the other ids.

Reply:
But it also doesn't really matter since these only check for equality.

Reply:
Yeah, irrelevant.